editor/backend/src/EditBuiltInCommand.ts

```diff
@@ -245,3 +245,267 @@ export class BasicManipulationCommand extends EditCommand implements BasicManipu
     this.iModel.updateEcefLocation(newEcefLocation);
   }
 }
+
+interface ElementGeometryCacheRequestProps {
+  id?: Id64String;
+}
+
+interface ElementGeometryCacheResponseProps {
+  status: BentleyStatus;
+  numGeom?: number;
+  numPart?: number;
+  numSolid?: number;
+  numSurface?: number;
+  numCurve?: number;
+  numOther?: number;
+}
+
+interface SubEntityGeometryResponseProps {
+  /** The face, edge, or vertex geometry */
+  geometry: ElementGeometryDataEntry;
+  /** The face or edge range box for the sub-entity geometry stored as 6 values for low/high */
+  range?: Float64Array;
+  /** Category id for geometry. */
+  category?: Id64String;
+  /** SubCategory id for geometry. */
+  subCategory?: Id64String;
+  /** Material id for geometry. */
+  material?: Id64String;
+  /** color of geometry. */
+  color?: ColorDefProps;
+  /** transparency of geometry. */
+  transparency?: number;
+  /** weight of geometry. */
+  weight?: number;
+}
+
+type SubEntityGeometryFunction = (info: SubEntityGeometryResponseProps) => void;
+type ClosestFaceFunction = (info: SubEntityLocationProps) => void;
+
+interface SubEntityGeometryRequestProps {
+  /** Sub-entity to return geometry for */
+  subEntity: SubEntityProps;
+  /** Callback for result */
+  onResult: SubEntityGeometryFunction;
+}
+
+interface ClosestFaceRequestProps {
+  /** Space point */
+  point: XYZProps;
+  /** Optional direction for choosing face from edge or vertex hit... */
+  direction?: XYZProps;
+  /** Callback for result */
+  onResult: ClosestFaceFunction;
+}
+
+interface ElementGeometryCacheOperationRequestProps {
+  /** Target element id, tool element ids supplied by operations between elements... */
+  id: Id64String;
+  /** Callback for result when element's geometry stream is requested in flatbuffer or graphic formats */
+  onGeometry?: ElementGeometryFunction;
+  onSubEntityGeometry?: SubEntityGeometryRequestProps;
+  onClosestFace?: ClosestFaceRequestProps;
+  onOffsetFaces?: OffsetFacesProps;
+}
+
+/** @alpha */
+export class SolidModelingCommand extends BasicManipulationCommand implements SolidModelingCommandIpc {
+  public static override commandId = editorBuiltInCmdIds.cmdSolidModeling;
+
+  private async updateElementGeometryCache(props: ElementGeometryCacheRequestProps): Promise<ElementGeometryCacheResponseProps> {
+    const requestContext = new BackendRequestContext();

```

Comment:
can't all this be done with plain old `await`s?

Reply:
Maybe? Was just following what we do for snapping, etc. in IModelDb.ts...

Reply:
He's passing a callback to an async native function - he needs to call `resolve` inside the callback, so has to create a Promise.